lib/Bio/EnsEMBL/DataCheck/Checks/FeaturePosition.pm

```diff
@@ -23,67 +23,113 @@ use strict;
+	  attrib_type USING (attrib_type_id) INNER JOIN
+	  coord_system USING (coord_system_id)
+    WHERE
+      attrib_type.code = "toplevel" AND
+      coord_system.species_id = $species_id;
   /;
 
-  my $helper = $dna_dba->dbc->sql_helper;
-  my %regions = %{$helper->execute_into_hash(-SQL => $sql_regions)};
-  my @feature_tables = ("peak", "regulatory_feature", "motif_feature", "external_feature", "mirna_target_feature");
-  my $feature_pos_sql;
-  my $region_length;
-
-  foreach my $feature_table (@feature_tables) {
-    my $desc_feature_ok = "No " . $feature_table . "s exceeding seq_region bounds";
-    my $feature_count = 0;
-    foreach my $region_id (keys %regions) {
-      $region_length = $regions{$region_id};
-      if ($feature_table eq "regulatory_feature") {
-        $feature_pos_sql = qq/
-          SELECT COUNT(${feature_table}_id)
-          FROM    $feature_table
-          WHERE   seq_region_id = $region_id
-          AND     ((seq_region_start - bound_start_length) <= 0
-          OR      (seq_region_end + bound_end_length) > $region_length)
-          /;
-
-      }
-      else {
-        $feature_pos_sql = qq/
-          SELECT COUNT(${feature_table}_id)
-          FROM    $feature_table
-          WHERE   seq_region_id= $region_id
-          AND     (seq_region_start <= 0 OR seq_region_end > $region_length)
-          /;
-      }
-      $feature_count += sql_count($self->dba, $feature_pos_sql);
-    }
-    is($feature_count, 0, $desc_feature_ok);
+  my $seq_regions = $dna_dba->dbc->sql_helper->execute_into_hash( -SQL => $sql );
+
+  return $seq_regions;
+}
+
+sub start_bound_check {
+  my ($self, $table) = @_;
+
+  my $desc = "No ${table}s beyond start of seq_region";
+
+  my $sql;
+  if ($table eq "regulatory_feature") {
+    $sql = qq/
+      SELECT COUNT(*)
+      FROM $table
+      WHERE seq_region_start - bound_start_length <= 0
+    /;
+  } else {
+    $sql = qq/
+      SELECT COUNT(*)
+      FROM $table
+      WHERE seq_region_start <= 0
+    /;
   }
+
+  is_rows_zero($self->dba, $sql, $desc);
 }
 
-1;
+sub end_bound_check {
+  my ($self, $table, $seq_regions) = @_;
+
+  my $desc = "No ${table}s beyond end of seq_region";
+
+  my $sql;
+  if ($table eq "regulatory_feature") {
+    $sql = qq/
+      SELECT
+        seq_region_id,
+        MAX(seq_region_end + bound_end_length) AS max_length
+      FROM $table
+      GROUP BY seq_region_id
+    /;
+  } else {
+    $sql = qq/
+      SELECT
+        seq_region_id,
+        MAX(seq_region_end) AS max_length
+      FROM $table
+      GROUP BY seq_region_id
+    /;
+  }
+
+  my $max_ends = $self->dba->dbc->sql_helper->execute_into_hash( -SQL => $sql );
+
+  my @out_of_bounds;
 
+  foreach my $sr_id (keys %$max_ends) {
+	if (exists $$seq_regions{$sr_id}) {
+	  if ($$max_ends{$sr_id} > $$seq_regions{$sr_id}) {
+	    push @out_of_bounds, $sr_id;
+	  }
+	}
+  }
+
+  is(scalar(@out_of_bounds), 0, $desc) ||
+    diag explain {'seq__region_ids' => \@out_of_bounds};
```

Comment:
did you want to distinguish `seq__region_ids` from `seq_region_ids` or is this a typo?

Reply:
Thanks for spotting, that was a typo.